### PR TITLE
Deployment export

### PR DESCRIPTION
**Pull request by okuchaiev**
Adding action to be able to export Neural Modules for inference in various formats

### CODE REVIEW

**Comment by okuchaiev** [comment]
@blisc can you please have a look at  collections/nemo_asr/nemo_asr/parts/jasper.py ?

File: nemo/nemo/backends/pytorch/actions.py [hunk 1]
```diff
@@ -97,7 +100,7 @@ def __get_top_sorted_modules_and_dataloader(self, hook):
           in DAG
 
         Returns:
-          list of modules with their call arguments and outputs, and dataset
```
Comment (blisc) [inline nemo/nemo/backends/pytorch/actions.py]: I disagree with the changed comments.
module should be pluralize into modules whenever necessary.

File: collections/nemo_asr/nemo_asr/parts/jasper.py [hunk 2]
```diff
@@ -129,11 +149,11 @@ def __init__(self, inplanes, planes, repeat=3, kernel_size=11, stride=1,
         self.separable = separable
         self.residual_mode = residual_mode
 
-        self.conv = nn.ModuleList()
         inplanes_loop = inplanes
+        conv = nn.ModuleList()
 
-        if tied:
```
Comment (blisc) [inline collections/nemo_asr/nemo_asr/parts/jasper.py]: I see that we are removing the weight sharing option. Can we then remove it from the class initializer as a parameter?

**Comment by blisc** [comment]
> @blisc can you please have a look at collections/nemo_asr/nemo_asr/parts/jasper.py ?

It seems to be ok for Jasper. Does the changes work for QuartzNet?

File: nemo/nemo/backends/pytorch/actions.py (hunk 1 above)
Comment (okuchaiev) [reply nemo/nemo/backends/pytorch/actions.py]: sorry, this was done by wrong docstring re-factoring. will fix back

File: nemo/nemo/backends/pytorch/actions.py [hunk 3]
```diff
@@ -374,7 +377,7 @@ def __initialize_amp(
             return optimizer
 
         if len(self.modules) < 1:
-            raise ValueError("There were no modules to initialize")
+            raise ValueError("There were no module to initialize")
```
Comment (blisc) [inline nemo/nemo/backends/pytorch/actions.py]: Should be plurarl.

File: nemo/nemo/backends/pytorch/actions.py [hunk 4]
```diff
@@ -971,8 +974,8 @@ def _check_tuples(list_of_tuples):
 
     def _get_all_modules(
             self, training_loop, callbacks, logging_callchain=None):
-        """Gets all neural modules that will be used by train() and eval() via
-        EvaluatorCallbacks. Saves all modules to self.modules
+        """Gets all neural module that will be used by train() and eval() via
```
Comment (blisc) [inline nemo/nemo/backends/pytorch/actions.py]: In this docstring, first two should be plural, last one depends on the name of the variable but I am guessing not.

File: nemo/nemo/backends/pytorch/actions.py [hunk 5]
```diff
@@ -987,7 +990,7 @@ def _get_all_modules(
                 for module in step[2]:
                     self.modules.add(module[0])
 
-        # Lastly, grab all eval modules
+        # Lastly, grab all eval module
```
Comment (blisc) [inline nemo/nemo/backends/pytorch/actions.py]: Should be plural.

File: nemo/nemo/backends/pytorch/actions.py [hunk 6]
```diff
@@ -1032,7 +1161,7 @@ def train(self,
                            NmTensor) and PtActions._check_all_tensors(
                 tensors_to_optimize)):
             # Parse graph into a topologically sorted sequence of neural
-            # modules' calls
+            # module' calls
```
Comment (blisc) [inline nemo/nemo/backends/pytorch/actions.py]: Should be plural.

File: nemo/nemo/backends/pytorch/actions.py [hunk 7]
```diff
@@ -1349,7 +1478,7 @@ def infer(self,
         """
 
         if checkpoint_dir:
-            # Find all modules that need to be restored
+            # Find all module that need to be restored
```
Comment (blisc) [inline nemo/nemo/backends/pytorch/actions.py]: Should be plural.

File: nemo/nemo/core/neural_factory.py [hunk 8]
```diff
@@ -148,7 +157,7 @@ def create_optimizer(
 
         Args:
             optimizer: Specifies which optimizer to use.
-            things_to_optimize: A list of neural modules or tensors to be
+            things_to_optimize: A list of neural module or tensors to be
```
Comment (blisc) [inline nemo/nemo/core/neural_factory.py]: Should be plural.

File: nemo/nemo/core/neural_factory.py [hunk 9]
```diff
@@ -230,14 +239,14 @@ class NeuralModuleFactory(object):
     _DEFAULT = None
 
     """
-    Neural Module Factory instance is used to create neural modules and
+    Neural Module Factory instance is used to create neural module and
```
Comment (blisc) [inline nemo/nemo/core/neural_factory.py]: Should be plural.

File: nemo/nemo/core/neural_factory.py [hunk 10]
```diff
@@ -230,14 +239,14 @@ class NeuralModuleFactory(object):
...
     trainers
 
     Args:
         backend (Backend): Currently only Backend.PyTorch is supported
         local_rank (int): Process rank. Should be set by distributed runner
         optimization_level (Optimization): Level of optimization to use. Will
-            be passed to neural modules and actions created by this factory.
+            be passed to neural module and actions created by this factory.
```
Comment (blisc) [inline nemo/nemo/core/neural_factory.py]: Should be plural.

File: nemo/nemo/backends/pytorch/actions.py [hunk 11]
```diff
@@ -997,6 +1000,132 @@ def _get_all_modules(
...
+            raise FileExistsError(f"Destination {output} already exists. "
+                                  f"Aborting export.")
+
+        input_names = list(module.input_ports.keys())
+        output_names = list(module.output_ports.keys())
+        dynamic_axes = {}
+
+        def __extract_dynamic_axes(port_name: str, ntype: NeuralType,
+                                   dynamic_axes: Dict):
+            if ntype.axis2type is not None and ntype.axis2type != {}:
```
Comment (redoctopus) [inline nemo/nemo/backends/pytorch/actions.py]: I think `if ntype.axis2type:` should take care of both cases (None and empty dict => False)

File: nemo/nemo/backends/pytorch/actions.py [hunk 12]
```diff
@@ -997,6 +1000,132 @@ def _get_all_modules(
...
+                for axis_id, axistype in ntype.axis2type.items():
+                    if issubclass(axistype.semantics, BatchTag) or issubclass(
+                            axistype.semantics, TimeTag):
+                        if port_name not in dynamic_axes:
+                            dynamic_axes[port_name] = []
```
Comment (redoctopus) [inline nemo/nemo/backends/pytorch/actions.py]: Any thoughts about using defaultdict for `dynamic_axes` so we can make this cleaner?

File: nemo/nemo/backends/pytorch/actions.py [hunk 13]
```diff
@@ -997,6 +1000,132 @@ def _get_all_modules(
...
+                                  opset_version=10)
+                # fn = output + ".readable"
+                # with open(fn, 'w') as f:
+                #     # Write human-readable graph representation to file as
+                #     well.
+                #     tempModel = onnx.load(output)
+                #     onnx.save(tempModel, output + ".copy")
+                #     onnx.checker.check_model(tempModel)
+                #     pgraph = onnx.helper.printable_graph(tempModel.graph)
+                #     f.write(pgraph)
```
Comment (redoctopus) [inline nemo/nemo/backends/pytorch/actions.py]: What's this? Looks useful--should it be an option? Are we coming back to this?

File: nemo/nemo/backends/pytorch/actions.py [hunk 14]
```diff
@@ -997,6 +1000,132 @@ def _get_all_modules(
...
+
+            elif d_format == DeploymentFormat.PYTORCH:
+                torch.save(module.state_dict(), output)
+                with open(output + ".json", 'w') as outfile:
+                    json.dump(local_parameters, outfile)
+
+            else:
+                raise NotImplemented(
+                    f"Not supported deployment format: {d_format}")
+        except:  # nopep8
```
Comment (redoctopus) [inline nemo/nemo/backends/pytorch/actions.py]: Does this fail silently?

File: nemo/nemo/backends/pytorch/actions.py (hunk 12 above)
Comment (okuchaiev) [reply nemo/nemo/backends/pytorch/actions.py]: can you give example?

Comment (redoctopus) [reply nemo/nemo/backends/pytorch/actions.py]: Such as:
```python
from collections import defaultdict
...
dynamic_axes = defaultdict(list)
# And then we can directly do:
dynamic_axes[port_name].append(axis_id)
```
The defaultdict just gets rid of the need for the check here.

File: collections/nemo_asr/nemo_asr/parts/jasper.py (hunk 2 above)
Comment (okuchaiev) [reply collections/nemo_asr/nemo_asr/parts/jasper.py]: done

File: nemo/nemo/backends/pytorch/actions.py (hunk 3 above)
Comment (okuchaiev) [reply nemo/nemo/backends/pytorch/actions.py]: done

File: nemo/nemo/backends/pytorch/actions.py (hunk 5 above)
Comment (okuchaiev) [reply nemo/nemo/backends/pytorch/actions.py]: done

File: nemo/nemo/backends/pytorch/actions.py (hunk 7 above)
Comment (okuchaiev) [reply nemo/nemo/backends/pytorch/actions.py]: done

File: nemo/nemo/core/neural_factory.py (hunk 9 above)
Comment (okuchaiev) [reply nemo/nemo/core/neural_factory.py]: done

File: nemo/nemo/core/neural_factory.py (hunk 10 above)
Comment (okuchaiev) [reply nemo/nemo/core/neural_factory.py]: done

File: nemo/nemo/backends/pytorch/actions.py (hunk 11 above)
Comment (okuchaiev) [reply nemo/nemo/backends/pytorch/actions.py]: done

File: nemo/nemo/backends/pytorch/actions.py (hunk 12 above)
Comment (okuchaiev) [reply nemo/nemo/backends/pytorch/actions.py]: changed

File: nemo/nemo/backends/pytorch/actions.py (hunk 14 above)
Comment (okuchaiev) [reply nemo/nemo/backends/pytorch/actions.py]: changed - now prints error

**Comment by okuchaiev** [comment]
can we merge this?

**Pull request merged**